ortools/sat/timetable_horizontal_edgefinding.cc

```diff
@@ -0,0 +1,290 @@
+      demands_(demands),
+      integer_trail_(model->GetOrCreate<IntegerTrail>()) {
+  task_types_.resize(num_tasks_);
+  profile_events_.reserve(4 * num_tasks_ + 1);
+}
+
+void HorizontallyElasticOverloadChecker::RegisterWith(
+    GenericLiteralWatcher* watcher) {
+  const int id = watcher->Register(this);
+  watcher->WatchUpperBound(capacity_, id);
+  helper_->WatchAllTasks(id, watcher);
+  for (int t = 0; t < num_tasks_; t++) {
+    watcher->WatchLowerBound(demands_->Demands()[t], id);
+  }
+  watcher->SetPropagatorPriority(id, 3);
+  watcher->NotifyThatPropagatorMayNotReachFixedPointInOnePass(id);
+}
+
+bool HorizontallyElasticOverloadChecker::Propagate() {
+  if (!helper_->SynchronizeAndSetTimeDirection(true)) return false;
+  if (!OverloadCheckerPass()) return false;
+  if (!helper_->SynchronizeAndSetTimeDirection(false)) return false;
+  if (!OverloadCheckerPass()) return false;
+  return true;
+}
+
+bool HorizontallyElasticOverloadChecker::OverloadCheckerPass() {
+  // Prepate the profile events which will be used during ScheduleTasks to
+  // dynamically compute the profile. The events are valid for the entire
+  // function and do not need to be recomputed. 
+  //
+  // TODO: This datastructure contains everything we need to compute the 
+  // "classic" profile used in Time-Tabling. 
+  profile_events_.clear();
+  for (int i = 0; i < num_tasks_; i++) {
+    if (helper_->IsPresent(i) && demands_->DemandMin(i) > 0) {
+      profile_events_.emplace_back(i, helper_->StartMin(i),
+                                   demands_->DemandMin(i),
+                                   ProfileEventType::kStartMin);
+      profile_events_.emplace_back(i, helper_->StartMax(i),
+                                   demands_->DemandMin(i),
+                                   ProfileEventType::kStartMax);
+      profile_events_.emplace_back(i, helper_->EndMin(i),
+                                   demands_->DemandMin(i),
+                                   ProfileEventType::kEndMin);
+      profile_events_.emplace_back(i, helper_->EndMax(i),
+                                   demands_->DemandMin(i),
+                                   ProfileEventType::kEndMax);
+    }
+  }
+  profile_events_.emplace_back(-1, kMaxIntegerValue, IntegerValue(0),
+                               ProfileEventType::kSentinel);
+  IncrementalSort(profile_events_.begin(), profile_events_.end());
+
+  // Iterate on all the windows [-inf, window_end] where window_end is the end
+  // max of a task.
+  IntegerValue window_end = kMinIntegerValue;
+  for (const ProfileEvent event : profile_events_) {
+    if (event.event_type != ProfileEventType::kEndMax ||
+        event.time <= window_end) {
+      continue;
+    }
+    window_end = event.time;
+
+    const IntegerValue capacity = integer_trail_->UpperBound(capacity_);
+    if (window_end < ScheduleTasks(window_end, capacity)) {
+      AddScheduleTaskReason(window_end);
+      return helper_->ReportConflict();
```

Comment:
If end_min is lower than the result of ScheduleTasks, can't you push a tighter end_min?

Reply:
Good question, I guess we could use `ScheduleTasks(window_end, capacity) - 1` but I would have to think a little more about it. 

That being said, I don't think propagator as it stands (compared to its extensions that update the start/end vars) is worth adding to OR-tools. I'd rather prefer to optimize the explanation of these versions.